token-tracker/src/truth-sources/token-grant.js

```diff
@@ -0,0 +1,142 @@
+/** @typedef { import("../lib/contract-helper").ContractInstance } ContractInstance*/
+
+import { ITruthSource } from "./truth-source.js"
+import { Contract } from "../lib/contract-helper.js"
+import { logger } from "../lib/winston.js"
+
+import { EthereumHelpers } from "@keep-network/tbtc.js"
+const { callWithRetry } = EthereumHelpers
+
+import { getPastEvents } from "../lib/ethereum-helper.js"
+import { dumpDataToFile } from "../lib/file-helper.js"
+
+import TokenGrantJSON from "@keep-network/keep-core/artifacts/TokenGrant.json"
+
+import Web3 from "web3"
+const { toBN } = Web3.utils
+
+import { resolveGrantee } from "../lib/owner-lookup.js"
+
+const TOKEN_GRANT_DUMP_PATH = "./tmp/token-grant.json"
+const TOKEN_GRANT_BALANCES_DUMP_PATH = "./tmp/token-grant-balances.json"
+
+export class TokenGrantTruthSource extends ITruthSource {
+  /** @property {ContractInstance} tokenGrant */
+
+  constructor(/** @type {Context} */ context, /** @type {Number}*/ finalBlock) {
+    super(context, finalBlock)
+  }
+
+  async initialize() {
+    const TokenGrant = new Contract(TokenGrantJSON, this.context.web3)
+
+    this.context.addContract("TokenGrant", TokenGrant)
+
+    this.tokenGrant = await this.context.getContract("TokenGrant").deployed()
+  }
+
+  /**
+   * @typedef {Object} TokenGrant
+   * @property {string} id
+   * @property {Address} grantee
+   * @property {BN} balance
+   */
+
+  /**
+   * Gets all token grants based on TokenGrantCreated event of TokenGrant contract.
+   * Returns details of each grant like ID, grantee and balance. The balance is
+   * calculated at the pre defined final block as total grant amount minus withdrawn
+   * and revoked tokens. The balance includes tokens that were staked, it's important
+   * to exclude the stake amount in TokenStaking source of truth to avoid value
+   * duplication.
+   * @return {TokenGrant[]} Token grant with ID, grantee and balance at final block.
+   */
+  async findTokenGrants() {
+    logger.info(
+      `looking for TokenGrantCreated events emitted from ${this.tokenGrant.options.address} ` +
+        `between blocks ${this.context.deploymentBlock} and ${this.finalBlock}`
+    )
+
+    /** @type {Array} */
+    const events = await getPastEvents(
+      this.context.web3,
+      this.tokenGrant,
+      "TokenGrantCreated",
+      this.context.deploymentBlock,
+      this.finalBlock
+    )
+    logger.info(`found ${events.length} token grant created events`)
+
+    /** @type {String[]} */
+    const tokenGrantIDs = events.map((event) => event.returnValues.id)
+
+    /** @type {TokenGrant[]} */
+    const tokenGrants = []
+    for (const id of tokenGrantIDs) {
+      const grant = await callWithRetry(
+        this.tokenGrant.methods.getGrant(id),
+        undefined,
+        undefined,
+        this.finalBlock
+      )
+
+      // Balance is the total grant amount minus revoked and withdrawn tokens. The
+      // value includes staked tokens.
```

Comment:
But it does not take into account slashed tokens... 😬

Reply:
What I think we can do:
- Take all `TokenGrantStaked` events for the given grant ID from `TokenGrant` contract.
- Take all `DepositRedelegated` events for the given grant ID from `TokenStakingEscrow` contract.
- Extract operator addresses.
- Take all `TokensSlashed` events for extracted operators, note `amount`.
- Take all `TokensSeized` events for extracted operators, note `amount`.
- When calculating `const balance` in the next line, `sub` the two noted amounts.

Reply:
What if we subtract the staked amount in `token-grant.js` and resolve the staked tokens amount from grants in `token-staking.js`?

Reply:
Resolving staked amounts from grants will be more complex unless I don't fully understand your proposition. The advantage of the current (and proposed) approach is that we don't have to care about the staking escrow used only for grants. It may happen that from the perspective of `TokenGrant`, the given amount is still staked while `TokenStaking` says the active stake is 0. This amount (and more!) can live in multiple slots in `TokenStakingEscrow`.

Reply:
> Take all DepositRedelegated events for the given grant ID from TokenStakingEscrow contract.

Do we need to bother about `TokenStakingEscrow`? We could check both `TokenStaking` and `OldTokenStaking` for the slash and seize events. Am I correct?

Reply:
Solution proposed in e8c7bb2d6c80fd7ddb693b9b8edd6bb2695bda33

Reply:
> > Take all DepositRedelegated events for the given grant ID from TokenStakingEscrow contract.
> 
> Do we need to bother about `TokenStakingEscrow`? We could check both `TokenStaking` and `OldTokenStaking` for the slash and seize events. Am I correct?

I haven't had a chance to review the proposed solution yet but the reason I mentioned `TokenStakingEscrow` is that the stake from a grant could be unstaked, tokens could be deposited into `TokenStakingEscrow`, and then redelegated to another operator. In this scenario, we need to inspect `TokensSlashed` and `TokensSeized` for two operator addresses.

Reply:
Covered in 26779ac6e262614345fe3792aff74cf523276cf1. I also included any withdrwan or revoked tokens.